### PR TITLE
Report error as structured data in fabric runner.

### PR DESCRIPTION
**Pull request by manasdk**
- Remote script action
```
$ st2 run default.data_gen hosts="st2master001" chars=1000 -a
+-----------------+-----------------------------+
| Property        | Value                       |
+-----------------+-----------------------------+
| id              | 54b6d5600640fd62c48bac2c    |
| context         | {                           |
|                 |     "user": "stanley"       |
|                 | }                           |
| parameters      | {                           |
|                 |     "chars": 1000,          |
|                 |     "hosts": "st2master001" |
|                 | }                           |
| status          | scheduled                   |
| start_timestamp | 2015-01-14T20:45:20.943512Z |
| result          |                             |
| action          | default.data_gen            |
| callback        |                             |
+-----------------+-----------------------------+
To get the results, execute:
    $ st2 execution get 54b6d5600640fd62c48bac2c
(virtualenv) (fabric-runner-output) ~/code/st2$ st2 execution get 54b6d5600640fd62c48bac2c
STATUS: failed
RESULT:
{
    "st2master001": {
        "failed": true,
        "traceback": "  File "/home/manas/code/st2/st2common/st2common/models/system/action.py", line 382, in _run_script
    self._execute_remote_command('mkdir %s' % self.remote_dir)
  File "/home/manas/code/st2/st2common/st2common/models/system/action.py", line 428, in _execute_remote_command
    output = action_method(command, combine_stderr=False, pty=False, quiet=True)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 639, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/operations.py", line 1042, in run
    shell_escape=shell_escape)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/operations.py", line 909, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/state.py", line 390, in default_channel
    chan = _open_session()
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/state.py", line 382, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 151, in __getitem__
    self.connect(key)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 143, in connect
    self[key] = connect(user, host, port, cache=self)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 533, in connect
    raise NetworkError('Name lookup failed for %s' % host, e)
",
        "succeeded": false,
        "error": "Name lookup failed for st2master001"
    }
}
```

- Remote command action
```
$ st2 run core.remote hosts="st2master001" cmd="date" -a
+-----------------+-----------------------------+
| Property        | Value                       |
+-----------------+-----------------------------+
| id              | 54b6d59f0640fd62c48bac2d    |
| context         | {                           |
|                 |     "user": "stanley"       |
|                 | }                           |
| parameters      | {                           |
|                 |     "cmd": "date",          |
|                 |     "hosts": "st2master001" |
|                 | }                           |
| status          | scheduled                   |
| start_timestamp | 2015-01-14T20:46:23.471093Z |
| result          |                             |
| action          | core.remote                 |
| callback        |                             |
+-----------------+-----------------------------+
To get the results, execute:
    $ st2 execution get 54b6d59f0640fd62c48bac2d
(virtualenv) (fabric-runner-output) ~/code/st2$ st2 execution get 54b6d59f0640fd62c48bac2d
STATUS: failed
RESULT:
{
    "st2master001": {
        "failed": true,
        "traceback": "  File "/home/manas/code/st2/st2common/st2common/models/system/action.py", line 335, in _run
    output = run(self.command, combine_stderr=False, pty=False, quiet=True)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 639, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/operations.py", line 1042, in run
    shell_escape=shell_escape)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/operations.py", line 909, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/state.py", line 390, in default_channel
    chan = _open_session()
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/state.py", line 382, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 151, in __getitem__
    self.connect(key)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 143, in connect
    self[key] = connect(user, host, port, cache=self)
  File "/home/manas/code/st2/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 533, in connect
    raise NetworkError('Name lookup failed for %s' % host, e)
",
        "succeeded": false,
        "error": "Name lookup failed for st2master001"
    }
}
```

### CODE REVIEW

File: st2common/st2common/models/system/action.py [hunk 1]
```diff
@@ -85,6 +84,20 @@ def _get_command_string(self, cmd, args):
...
+        Prepares a structured error result based on the exception.
+
+        :type e: ``Exception``
+
+        :rtype: ``dict``
+        """
+        result = {}
+        result['failed'] = True
+        result['succeeded'] = False
+        result['exception'] = str(e)
```
Comment (lakshmi-kannan) [inline st2common/st2common/models/system/action.py]: heh. tb.format_exc()

**Comment by lakshmi-kannan** [comment]
Travis and tb.format_exc(). LGTM otherwise.

**Comment by manasdk** [comment]
Travis unhappy because of ... 
Installing test-requirements.txt...
  Hash of the package https://pypi.python.org/packages/source/d/docutils/docutils-0.12.tar.gz#md5=4622263b62c5c771c03502afa3157768 (from https://pypi.python.org/simple/docutils/) (0618edf58e58000edc3102b5988ad163) doesn't match the expected hash 4622263b62c5c771c03502afa3157768!
  Bad md5 hash for package https://pypi.python.org/packages/source/d/docutils/docutils-0.12.tar.gz#md5=4622263b62c5c771c03502afa3157768 (from https://pypi.python.org/simple/docutils/)

I hope the package on pypi gets fixed.

File: st2common/st2common/models/system/action.py (hunk 1 above)
Comment (manasdk) [inline st2common/st2common/models/system/action.py]: I think that putting the traceback in the result makes it hard to read the result. Also, less human friendly. Thoughts?

**Comment by manasdk** [comment]
Ignoring the travis failure as the issue is out of our control. The pypi docutils package md5 needs to be fixed.

File: st2common/st2common/models/system/action.py [hunk 2]
```diff
@@ -313,15 +330,19 @@ def _get_action_method(self):
...
+            with shell_env(**self.env_vars), settings(command_timeout=self.timeout):
+                output = run(self.command, combine_stderr=False, pty=False, quiet=True)
+            result = {
+                'stdout': output.stdout,
+                'stderr': output.stderr,
+                'return_code': output.return_code,
+                'succeeded': output.succeeded,
+                'failed': output.failed
+            }
+        except Exception:
```
Comment (Kami) [inline st2common/st2common/models/system/action.py]: Is there a reason why we don't do the same in `_sudo` method?

**Pull request merged**